x/ccv/consumer/keeper/keeper.go

```diff
@@ -206,50 +206,95 @@ func (k Keeper) DeletePendingChanges(ctx sdk.Context) {
+	k.cdc.MustUnmarshal(bz, &validatorSetChangeIDs)
 
-		stop := cb(seq, timeNs)
-		if stop {
-			break
-		}
+	return validatorSetChangeIDs.Ids
+}
+
+// SetVSCPacketQueueTimeSlice sets a specific VSC packet queue timeslice.
+func (k Keeper) SetVSCPacketQueueTimeSlice(ctx sdk.Context, timestamp time.Time, vscIDs ...uint64) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshal(&types.ValidatorSetChangeIDs{Ids: vscIDs})
+	store.Set(types.VSCPacketQueueKey(timestamp), bz)
+}
+
+// InsertVSCPacketQueue inserts the ID of a VSC packet to the appropriate timeslice
+// in the VSCPacketQueue.
+func (k Keeper) InsertVSCPacketQueue(ctx sdk.Context, vscID uint64, maturityTime time.Time) {
+	timeSlice := k.GetVSCPacketQueueTimeSlice(ctx, maturityTime)
+	if len(timeSlice) == 0 {
+		k.SetVSCPacketQueueTimeSlice(ctx, maturityTime, vscID)
+	} else {
+		timeSlice = append(timeSlice, vscID)
+		k.SetVSCPacketQueueTimeSlice(ctx, maturityTime, timeSlice...)
 	}
 }
 
-// SetPacketMaturityTime sets the maturity time for a given received VSC packet id
-func (k Keeper) SetPacketMaturityTime(ctx sdk.Context, vscId, maturityTime uint64) {
+// VSCPacketQueueIterator returns all the VSC packet queue timeslices from time 0 until endTime.
+func (k Keeper) VSCPacketQueueIterator(ctx sdk.Context, endTime time.Time) sdk.Iterator {
 	store := ctx.KVStore(k.storeKey)
-	timeBytes := make([]byte, 8)
-	binary.BigEndian.PutUint64(timeBytes, maturityTime)
-	store.Set(types.PacketMaturityTimeKey(vscId), timeBytes)
+	return store.Iterator([]byte{types.VSCPacketQueueBytePrefix},
+		sdk.InclusiveEndBytes(types.VSCPacketQueueKey(endTime)))
 }
 
-// GetPacketMaturityTime gets the maturity time for a given received VSC packet id
-func (k Keeper) GetPacketMaturityTime(ctx sdk.Context, vscId uint64) uint64 {
+// DequeueAllMatureVSCPacketQueue returns a concatenated list of all IDs of matured VSC packets
+// and deletes the timeslices from the VSC packet queue.
+func (k Keeper) DequeueAllMatureVSCPacketQueue(ctx sdk.Context) (vscIDs []uint64) {
 	store := ctx.KVStore(k.storeKey)
-	bz := store.Get(types.PacketMaturityTimeKey(vscId))
-	if bz == nil {
-		return 0
+
+	// gets an iterator for all timeslices from time 0 until the current Blockheader time
+	iterator := k.VSCPacketQueueIterator(ctx, ctx.BlockHeader().Time)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		timeslice := types.ValidatorSetChangeIDs{}
+		value := iterator.Value()
+		k.cdc.MustUnmarshal(value, &timeslice)
+
+		vscIDs = append(vscIDs, timeslice.Ids...)
+
+		store.Delete(iterator.Key())
 	}
-	return binary.BigEndian.Uint64(bz)
+
+	return vscIDs
 }
 
-// DeletePacketMaturityTimes deletes the packet maturity time for given received VSC packet ids
-func (k Keeper) DeletePacketMaturityTimes(ctx sdk.Context, vscIds ...uint64) {
+// GetAllVSCPacketMaturityTimes returns the maturity times of all received VSC packets.
+//
+// Note: The order of iteration is irrelevant as it returns all maturity times.
```

Comment:
Is there an expectation about the order of iteration expressed somewhere else in the code?

If not, it is confusing why a comment about order of iteration was necessary. The function returns a slice of all times, the slice can be sorted  by `MaturingVSCPacket.MaturityTime` it the caller needs specific sort ordering